### DISCUSSION TITLE
remember and restore temperature in "Warmwasser bereiten"

### DISCUSSION THREAD

**Discussion by svde** [Trunks1982/Daikin-Rotex-HPSU-CAN#32] [Ideas]
At the moment the "Warmwasser bereiten" function sets the temperature back to 45C. It would be nice to store the current set t_ww_soll value and set that back after the short bump to 70C to trigger a DHW cycle.

**Comment by Trunks1982** [comment]
Jap fände ich auch cool aber mit esphome hab ich es noch nicht hinbekommen.Schätze es muss einen extra sensor geben der geschrieben wird und dann wieder abgefragt wird.Chatgpt konnte mir da nicht wirklich helfen🤣

**Comment by svde** [comment]
Tried this:

Declare a global variable to store the value of t_ww_soll
```
globals:
  - id: backup_dhw
    type: int
```

and then:
```
button:
  - platform: template
    name: Warmwasser bereiten
    id: ww_button
    icon: "mdi:water-boiler"
    on_press:
      then:
        - lambda: |-
           id(backup_dhw) = int(id(t_ww_soll) *10);
        - canbus.send:
            can_id: 0x680
            data: [ 0x30, 0x00, 0x13, 0x02, 0xBC, 0x00, 0x00 ] # 70 Grad Grundeinstellung
        - delay: 10s
        - canbus.send:
            can_id: 0x680
            data: !lambda |-
              uint16_t temperature = (uint16_t)(backup_dhw); // Convert to int16be
              uint8_t high_byte = temperature >> 8;
              uint8_t low_byte = temperature & 0xFF;
              return { 0x30, 0x00, 0x13, high_byte, low_byte, 0x00, 0x00 };
        - logger.log: CAN Nachricht wurde gesendet
#       - globals.set:
#           id: backup_dhw
#           value: !lambda return int(id(t_ww_soll).state *10);
```

but it doesn't compile due to the fact that t_ww_soll is a float which is incompatible with uint16_t.

**Reply by Trunks1982** [reply]
Heisst also unit16 kampatibel machen oder wie meinst?

**Reply by svde** [reply]
Yes:
```
Compiling .pioenvs/esphome-70041dd36608/src/main.cpp.o
/config/esphome-70041dd36608.yaml: In lambda function:
/config/esphome-70041dd36608.yaml:250:21: error: expected initializer before '->' token
             int id(backup_dhw) = (id(t_ww_soll) *10);
                     ^~
/config/esphome-70041dd36608.yaml: In lambda function:
/config/esphome-70041dd36608.yaml:261:51: error: cast from 'esphome::globals::GlobalsComponent<int>*' to 'uint16_t' {aka 'short unsigned int'} loses precision [-fpermissive]
               uint16_t temperature = (uint16_t)(backup_dhw); // Convert to int16be
                                                   ^
*** [.pioenvs/esphome-70041dd36608/src/main.cpp.o] Error 1
========================== [FAILED] Took 2.79 seconds ==========================
```

**Reply by Trunks1982** [reply]
Mhm

**Comment by Trunks1982** [comment]
So bekomme ich es kompiliert.Schaltet auf 70Grad aber dann nicht wieder zurück. 




globals:
  - id: backup_dhw
    type: int  



- platform: template
    name: Warmwasser bereiten1
    id: ww_button1
    icon: "mdi:water-boiler"
    on_press:
      then:
        - lambda: |-
            int backup_dhw = (int(id(t_ww_soll)) * 10);
            // Setze den Wert der globalen Variable backup_dhw auf das 10-fache des aktuellen Wertes der globalen Variable t_ww_soll
        - canbus.send:
            can_id: 0x680
            data: [ 0x30, 0x00, 0x13, 0x02, 0xBC, 0x00, 0x00 ] # 70 Grad Grundeinstellung
        - delay: 10s
        - canbus.send:
            can_id: 0x680
            data: !lambda |-
              uint16_t temperature = (uint16_t)(int(id(backup_dhw))); // Convert to int16be
              uint8_t high_byte = temperature >> 8;
              uint8_t low_byte = temperature & 0xFF;
              return { 0x30, 0x00, 0x13, high_byte, low_byte, 0x00, 0x00 };
        - logger.log: CAN Nachricht wurde gesendet
        - globals.set:
            id: backup_dhw
            value: !lambda return int(id(t_ww_soll).state *10);

**Comment by Trunks1982** [comment]
der code funktioniert:-)


'''
- platform: template
    name: Warmwasser bereiten1
    id: ww_button1
    icon: "mdi:water-boiler"
    on_press:
      then:
        - globals.set:
            id: backup_dhw
            value: !lambda return int(id(t_ww_soll).state * 10);
        - canbus.send:
            can_id: 0x680
            data: [0x30, 0x00, 0x13, 0x02, 0xBC, 0x00, 0x00] # 70 Grad Grundeinstellung
        - delay: 10s
        - canbus.send:
            can_id: 0x680
            data: !lambda |-
              uint16_t temperature = (uint16_t)id(backup_dhw); // Convert to int16be
              uint8_t high_byte = temperature >> 8;
              uint8_t low_byte = temperature & 0xFF;
              return {0x30, 0x00, 0x13, high_byte, low_byte, 0x00, 0x00};
        - logger.log: CAN Nachricht wurde gesendet
'''

**Comment by svde** [comment]
Nice! Will test tomorrow morning.

Good teamwork.

**Reply by Trunks1982** [reply]
Yes

**Reply by Trunks1982** [reply]
Hast chatgpt benutzt ne?😉

**Reply by svde** [reply]
No chatgpt for this one. Just reading the esphome documentation and searching on the homeassistent forum.  Commit is here: https://github.com/svde/Daikin-Rotex-HPSU-CAN/commit/7d5b46c75a89bdc5c9aebd96aff1c54a5093228d

**Reply by Trunks1982** [reply]
Das komische diesen code hatte ich ähnlich von chatgpt🤣 jnd habe deinen durch verbessern lassen

**Reply by svde** [reply]
What did chatgpt come up with?

**Reply by Trunks1982** [reply]
Deinen code hatte ich schon einmal ähnlich aus chatgpt.

**Comment by svde** [comment]
Tested, it works. Thank you.
